README
This is a tool to generate nice sequence (flow) diagrams from
textual descriptions, written in Haskell.

This tool generates a diagram _description_, which could be made into
a nice picture with help of Graphviz (www.graphviz.org)

To try it out, run:
  runhaskell flow2dot.hs sample.flow | dot -T png -o sample.png
and view "sample.png" with your favorite picture viewer. If you
dont get a nice picture and get something else (for example, ugly
segfault from dot), try upgrading to latest Graphviz (2.12 or later)

If you want to use national alphabets, make sure that your .flow files
are encoded in UTF-8. If you want to tweak the output - read Dot manual
and use it for scaling, colors, pagination etc.

Latest version could be obtained via:
  darcs pull http://adept.linux.kiev.ua:8080/repos/flow2dot/

License     :  BSD-style (see the file LICENSE)
Send patches to dastapov@gmail.com (using "darcs send")

Thanks to Cale, quicksilver and roconnor from #haskell for
suggestions on how to modularize this. Thanks to Dema from
haskell@conference.jabber.ru for win32 testing. Gwern0 helped
to adapt this to GHC 6.8.2

Known issues
============

* Dot < 2.12 will most likely segfault on files generated by flow2dot

* If your version of dot complains about "Error: lost n1 n2 edge", try using
flow2dot-fix-dot-lost-edges in place of dot:

   runhaskell flow2dot your.flow | flow2dot-fix-dot-lost-edges -T png -o your.png

Trick is to remove 'constraint="false"' from all edges that dot reports as lost.



ChangeLog
tagged 0.7

  * Added Order directive (see sample.flow for usage)
  * Fix: identifier could not be empty
  * Removing useles import
  * Preserve UTF8 chars in "reflow"
  * Fix to preserve UTF8 chars in "showDot" (again)
  * Imported Dot.hs from dotgen 0.4.1
  * Fixed "lost edge" errors from graphviz

tagged 0.6.1

  * Forgot to add Text.Dot local override to cabal. Thnx to dons for spotting it

tagged 0.6

  * Fixed repo URL in README
  * Made sources compilable with QuickCheck2
  * Dotgen kills UTF-8 in graph attributes. Incorporated patched version of dotgen-0.2 into flow2dot until it will be fixed upstream
  * Added quickCheck.hs and rebuild.sh

tagged 0.5.1

  * Moved all flow diagram processing to separate module (exposed from this package)

tagged 0.4
  * Documentation fixes
  * Switched to `dotgen` package for all graphviz generation needs.
	    Dropped support of preformatted strings in flow files.

tagged 0.3.1
  * Fixed links to repo and docs

tagged 0.3
  * Ditched Text.UTF8 in favor of utf8-string library
  * Quotes inside messages are now properly escaped for Dot. Sample updated accordingly.

tagged 0.2.1
  * Version bump to 0.2.1 - ready for GHC 6.8.2
  * -Wall
  * +LANGUAGE pragmas

tagged 0.2
  * Some minor clarifications in docs
  * Dropped regex-based parser in favor of parsec-based due to issues with Unicode
  * Added QuickCheck to cabal, added LICENSE to Dot.hs
  * Added QuickCheck for Flow parser/pretty-printer
  * Moved graph generation into separate module
  * Added UTF8 module
  * cabalization

tagged 0.1
  * initial version


Text/Dot.hs
-- |
-- Module: Text.Dot
-- Copyright: Andy Gill
-- License: BSD3
--
-- Maintainer: Andy Gill <andygill@ku.edu>
-- Stability: unstable
-- Portability: portable
--
-- This module provides a simple interface for building .dot graph files, for input into the dot and graphviz tools. 
-- It includes a monadic interface for building graphs.

module Text.Dot 
	( 
	  -- * Dot
	  Dot		-- abstract
	  -- * Nodes
	, node
	, NodeId	-- abstract
	, userNodeId
	, userNode
	  -- * Edges
	, edge
        , edge'
	, (.->.)
	  -- * Showing a graph
	, showDot
	  -- * Other combinators
	, scope
	, attribute
	, share
	, same
	, cluster
	  -- * Simple netlist generation
	, netlistGraph
	) where

import Data.Char
import qualified Data.Map as M
import qualified Data.Set as S

data DotGraph = DotGraph [GraphElement]

data NodeId = NodeId String
	    | UserNodeId Int

instance Show NodeId where
  show (NodeId str) = str
  show (UserNodeId i) 
	| i < 0     = "u_" ++ show (negate i)
	| otherwise = "u" ++ show i

data GraphElement = GraphAttribute String String
		  | GraphNode NodeId        [(String,String)]
		  | GraphEdge NodeId NodeId [(String,String)]
		  | GraphEdge' NodeId (Maybe String) NodeId (Maybe String) [(String,String)]
		  | Scope           [GraphElement]
		  | SubGraph NodeId [GraphElement]

data Dot a = Dot { unDot :: Int -> ([GraphElement],Int,a) }

instance Monad Dot where
  return a = Dot $ \ uq -> ([],uq,a)
  m >>= k  = Dot $ \ uq -> case unDot m uq of
			   (g1,uq',r) -> case unDot (k r) uq' of
					   (g2,uq2,r2) -> (g1 ++ g2,uq2,r2)

-- | 'node' takes a list of attributes, generates a new node, and gives a 'NodeId'.
node      :: [(String,String)] -> Dot NodeId
node attrs = Dot $ \ uq -> let nid = NodeId $ "n" ++ show uq 
			  in ( [ GraphNode nid attrs ],succ uq,nid)


-- | 'userNodeId' allows a user to use their own (Int-based) node id's, without needing to remap them.
userNodeId :: Int -> NodeId
userNodeId i = UserNodeId i

-- | 'userNode' takes a NodeId, and adds some attributes to that node. 
userNode :: NodeId -> [(String,String)] -> Dot ()
userNode nId attrs = Dot $ \ uq -> ( [GraphNode nId attrs ],uq,())

-- | 'edge' generates an edge between two 'NodeId's, with attributes.
edge      :: NodeId -> NodeId -> [(String,String)] -> Dot ()
edge  from to attrs = Dot (\ uq -> ( [ GraphEdge from to attrs ],uq,()))

-- | 'edge' generates an edge between two 'NodeId's, with optional node sub-labels, and attributes.
edge'      :: NodeId -> Maybe String -> NodeId -> Maybe String -> [(String,String)] -> Dot ()
edge'  from optF to optT attrs = Dot (\ uq -> ( [ GraphEdge' from optF to optT attrs ],uq,()))

-- | '.->.' generates an edge between two 'NodeId's.
(.->.) from to = edge from to []

-- | 'scope' groups a subgraph together; in dot these are the subgraphs inside "{" and "}".
scope     :: Dot a -> Dot a
scope (Dot fn) = Dot (\ uq -> case fn uq of
			      ( elems,uq',a) -> ([Scope elems],uq',a))

-- | 'share' is when a set of nodes share specific attributes. Usually used for layout tweaking.
share :: [(String,String)] -> [NodeId] -> Dot ()
share attrs nodeids = Dot $ \ uq -> 
      ( [ Scope ( [ GraphAttribute name val | (name,val) <- attrs]
	       ++ [ GraphNode nodeid [] | nodeid <- nodeids ]
	       ) 
        ], uq, ())

-- | 'same' provides a combinator for a common pattern; a set of 'NodeId's with the same rank.
same :: [NodeId] -> Dot ()
same = share [("rank","same")]


-- | 'cluster' builds an explicit, internally named subgraph (called cluster). 
cluster :: Dot a -> Dot (NodeId,a)
cluster (Dot fn) = Dot (\ uq -> 
		let cid = NodeId $ "cluster_" ++ show uq 
		in case fn (succ uq) of
		    (elems,uq',a) -> ([SubGraph cid elems],uq',(cid,a)))

-- | 'attribute' gives a attribute to the current scope.
attribute :: (String,String) -> Dot ()
attribute (name,val) = Dot (\ uq -> ( [  GraphAttribute name val ],uq,()))

-- 'showDot' renders a dot graph as a 'String'.
showDot :: Dot a -> String
showDot (Dot dm) = case dm 0 of
		    (elems,_,_) -> "digraph G {\n" ++ unlines (map showGraphElement elems) ++ "\n}\n"

showGraphElement (GraphAttribute name val) = showAttr (name,val) ++ ";"
showGraphElement (GraphNode nid attrs)           = show nid ++ showAttrs attrs ++ ";"
showGraphElement (GraphEdge from to attrs) = show from ++ " -> " ++ show to ++  showAttrs attrs ++ ";"
showGraphElement (GraphEdge' from optF to optT attrs) = showName from optF ++ " -> " ++ showName to optT ++  showAttrs attrs ++ ";"
    where showName n Nothing = show n
          showName n (Just t) = show n ++ ":" ++ t
showGraphElement (Scope elems) = "{\n" ++ unlines (map showGraphElement elems) ++ "\n}"
showGraphElement (SubGraph nid elems) = "subgraph " ++ show nid ++ " {\n" ++ unlines (map showGraphElement elems) ++ "\n}"

showAttrs [] = ""
showAttrs xs = "[" ++ showAttrs' xs ++ "]"
    where
	-- never empty list
	showAttrs' [a]    = showAttr a
	showAttrs' (a:as) = showAttr a ++ "," ++ showAttrs' as

showAttr (name,val) = name ++ "=\""   ++ foldr showsDotChar "" val ++ "\""

showsDotChar '"'  = ("\\\"" ++)
showsDotChar '\\' = ("\\\\" ++)
showsDotChar x    | ord x < 32 = showLitChar x
                  | otherwise  = ([x]++)


-- | 'netlistGraph' generates a simple graph from a netlist.
netlistGraph :: (Ord a) 
          => (b -> [(String,String)])   -- ^ Attributes for each node
          -> (b -> [a])                 -- ^ Out edges leaving each node
          -> [(a,b)] 			-- ^ The netlist
	  -> Dot ()
netlistGraph attrFn outFn assocs = do
    let nodes = S.fromList $ [ a | (a,_) <- assocs ]
    let outs  = S.fromList $ [ o | (_,b) <- assocs
				 , o <- outFn b 
			     ]
    nodeTab <- sequence [ do nd <- node (attrFn b)
                             return (a,nd)
                        | (a,b) <- assocs ]
    otherTab <- sequence [ do nd <- node []
                              return (o,nd)
                         | o <- S.toList outs
                         , o `S.notMember` nodes
                         ]
    let fm = M.fromList (nodeTab ++ otherTab)
    sequence_ [ (fm M.! src) .->. (fm M.! dst)
              | (dst,b) <- assocs
              , src     <- outFn b
              ]
    return ()



flow2dot.hs
-----------------------------------------------------------------------------
-- |
-- Name        :  Flow2Dot
-- Copyright   :  (c) Dmitry Astapov, 2007-2009
-- License     :  BSD-style (see the file LICENSE)
--
-- Maintainer  :  Dmitry Astapov <dastapov@gmail.com>
-- Stability   :  beta
-- Portability :  portable
--
-----------------------------------------------------------------------------
module Main where

import Text.FlowDiagram
import System.IO.UTF8 (putStrLn)
import Prelude hiding (putStrLn)
import System (getArgs)

main :: IO ()
main = do
  args <- getArgs
  case args of
       [fname] -> process fname
       _ -> do print "Usage: flow2dot file.flow > file.dot"


-- | Process a .flow file and output generated .dot diagram
process :: FilePath -> IO ()
process fname = do
  flow <- parseFlowFromFile fname
  putStrLn $ flow2dot flow



quickCheck.hs
#!/usr/bin/runghc
{-

 Copyright (c) 2006 Dmitry Astapov <dastapov@gmail.com>
 Copyright (c) 2005 Lennart Kolmodin <kolmodin@dtek.chalmers.se>

 Permission to use, copy, modify, and distribute this software for any
 purpose with or without fee is hereby granted, provided that the above
 copyright notice and this permission notice appear in all copies.

 THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-}

{-
Inspired by similar script on John Hughes homepage
http://www.cs.chalmers.se/~rjmh/QuickCheck/quickcheck

This program looks through the current directory and searches for
haskell source files (.hs and .lhs).
In these files it looks for QuickCheck properties and
invokes ghci on them.

It's not tested that well, patches are welcome. :)
-}


module Main where

import Control.Monad

import Data.List

import System.Cmd
import System.Environment (getArgs)
import System.Exit
import System.Directory
import System.Posix.Files

-- options passed to ghci
opts = [ "-fglasgow-exts"  -- enable glasgow extensions
       , "-v0"             -- make ghci less verbose
       , "-w"              -- ignore all warnings
       ]

-- put the ghci of your choice here
ghci = "ghci"

test_opts = "TestOptions { no_of_tests = 100, length_of_tests = 5, debug_tests = False}"

main :: IO ()
main = do
  args <- getArgs
  let root = if null args then "." else head args
  files <- walkFiles root
  let hs_files = filter (not . exclude) files
  script <- liftM (unlines.concat) $ mapM prepare hs_files
  process opts script


-- | Determine if we should exclude the file from QuickCheck testing.
-- It has to be a haskell file (.hs or .lhs), and not start with "_darcs".
-- TODO: would it be more natural to say which files we should include instead?
exclude :: FilePath -> Bool
exclude fs =
    "_darcs" `isPrefixOf` fs || (not $ any (`isSuffixOf` fs) [".hs", ".lhs"])


-- | Return all files in a directory and its subdirectories.
-- Does not follow links or anything fancy like that.
walkFiles :: FilePath -> IO [FilePath]
walkFiles fp = do
  paths <- getDirectoryContents fp `catch` \_ -> return []
  let paths' = filter (`notElem` [".", ".."]) paths
  let loop files dirs [] = return (files, dirs)
      loop files dirs (x:xs) = do
        fs <- getFileStatus x
	case () of
	  _ | isRegularFile fs -> loop (x:files) dirs xs
	  _ | isDirectory fs   -> loop files (x:dirs) xs
	  _ | otherwise        -> loop files dirs xs -- ignore all other
  (files, dirs) <- loop [] [] $ map (fp </>) paths'
  rec <- concatMapM walkFiles dirs
  return (files ++ rec)
  where
  concatMapM f xs = liftM concat $ mapM f xs
  -- quick'n'dirty implementation of </>
  -- it joins two paths
  (</>) ""  b = b
  (</>) "." b = b
  (</>) a   b = a ++ '/':b


-- | Returns a list of properties defined in a file.
-- Ignore files it could not read and prints a warning.
-- TODO: We could use more fancy stuff for this. The current way is
--       a bit barbaric.
findProperties :: FilePath -> IO [String]
findProperties file_name = do
  raw_lines <- liftM lines (readFile file_name)
		`catch` \_ -> do
		putStrLn $ " !!! Could not read " ++ file_name
		return []
  let hs_lines | ".lhs" `isSuffixOf` file_name = map unlit raw_lines
               | otherwise		       = raw_lines
      props = nub $ filter ("prop_" `isPrefixOf`) $
  			map (fst . head . lex) hs_lines
  return props 
  where
  unlit ('>':' ':xs) = xs
  unlit xs           = xs

-- | Prepare a script for ghci which will run
-- all properties in given file
prepare :: FilePath -> IO [String]
prepare file_name = do
  props <- findProperties file_name
  return $ if (not . null $ props) 
           then [ ":l " ++ file_name
                , ":m + Test.QuickCheck Test.QuickCheck.Batch"
                , concat [ "runTests \""
                         , file_name 
                         , "\" (" ++ test_opts ++ ") [ " 
                         , concat (intersperse ", " $ map ("run "++) props) 
                         , " ]"
                         ]
                ]
           else []

-- | Run all the properties in a file through ghci.
-- TODO: What if we don't have write permission in the current directory?
--       It's better to create a proper temporary file than a fix one.
process :: [String] -> String -> IO ()
process options script = do
  writeFile "qc_in" script
  ec <- system (concat $ intersperse " " $ ghci : options ++ ["<qc_in"])
  case ec of
    ExitFailure i -> putStrLn $ "ghci exited with error " ++ show i
    _             -> return ()
  removeFile "qc_in"
  return ()



Text/FlowDiagram.hs
{-|
Converts flow diagrams to the Graphviz (Dot) files for subsequent rendering
into nice pictures.
-}
module Text.FlowDiagram ( flow2dot
                        , parseFlow
                        , parseFlowFromFile
                        , showFlow
                        , Flow(..)
                        ) where

import qualified Text.Dot as D
import Control.Monad.State (StateT, evalStateT, gets, modify, lift)
import qualified Data.Map as M (Map, empty, lookup, insert)
import Data.List (intercalate, unfoldr, splitAt, findIndex)
import Prelude hiding (readFile)
import System.IO.UTF8 (readFile)
import Data.Char (isSpace)
import Test.QuickCheck
import Control.Monad (liftM, liftM2, liftM3)
import Text.ParserCombinators.Parsec hiding (State)

{-
Idea: In order to draw sequence (flow) diagram using graphviz we can use directed layout (dot) to
generate "skeleton" of the diagram and draw message lines and action boxes

Diagram could look like this:
strict digraph SeqDiagram
{
  { // Those are swimline heads
    rank=same
    actor [label="Some actor"];
    system [label="Some system"];
  }
  { //tier1
    rank=same
    node[style=invis,shape=point];
    tier1; // this is an "anchor" for 1st diagram tier
    actor1; // this is a 1st point on "actor" swimline
    system1; // this is a 1st point on "system" swimline
  }
  { //tier2
    rank=same
    node[style=invis,shape=point];
    tier2; // anchor for 2nd diagram tier
    actor2; // this is a 2nd point on "actor" swimline
    system2; // this is a 2nd point on "system" swimline
  }
  // Main body

  // Tiers ordering. We link "anchor" nodes and Dot will do the rest
  tier1 -> tier2;

  // Actual messages. Note the "constraint=false"
  actor1 -> system1[label="xxx", constraint=false];
  system2 -> actor2[label="yyy", constraint=false];
}
-}

-- | Flow could include messages and actions, one item per source line
data Flow = Msg String String String
          -- ^ Message (from, to, message text). Syntax in the source file: @from -> to: message text@
          | Action String String
          -- ^ Action (actor, message text). Syntax in the source file: @actor: message text@
          | Order [String]
          -- ^ Tries to put swimlines in the specified order. Syntax: @order swimline1 swimline2 ...@
            deriving (Eq,Show)

-- | State of the diagram builder
data DiagS = DiagS { swimlines::M.Map String D.NodeId
                   -- ^ name of the swimline, ID of the last node on it
                   , numTier :: Int
                   -- ^ number of the next diagram tier
                   , headings :: [D.NodeId]
                   -- ^ IDs of all "swimline start" nodes
                   }

type Diagram = StateT DiagS D.Dot

-- | 'flow2dot' take a list of flow diagram items (`Flow') and converts them to Graphviz code
flow2dot :: [Flow] -> String
flow2dot flow = 
  ("strict "++) $ D.showDot $ evalStateT (flow2dot' flow) (DiagS M.empty 1 [])
    -- NB: "strict" is VERY important here
    -- Without it, "dot" segfaults while rendering diagram (dot 2.12)

flow2dot' :: [Flow] -> Diagram ()
flow2dot' flow = do
  let order = case [ ns | Order ns <- flow ] of
                []     -> Nothing
                -- Only the first Order directive would be taken into account
                (ns:_) -> Just ns
  mapM_ (flowElement2dot order) flow
  hs <- gets headings
  same hs

flowElement2dot :: Maybe [String] -> Flow -> Diagram ()
-- Make a graph block where swimline nodes for the current tier will be put.
-- Populate tier with "tier anchor" node
-- Generate nodes for message/action on all required swimlines
-- Connect generated nodes, if necessary
-- Connect tier to previous, which will ensure that tiers are ordered properly
flowElement2dot _ (Action actor message) = do
  tier <- invisNode
  l <- mkLabel message
  a <- node [("style","filled"),("shape","plaintext"),("label",l)]
  same [tier,a]
  connectToPrev actor a
  connectToPrev "___tier" tier
  incTier

flowElement2dot order (Msg from to message) = do
  tier <- invisNode
  f    <- invisNode
  t    <- invisNode
  same [f,t,tier]

  l <- mkLabel message

  connectToPrev from f
  connectToPrev to t
  connectToPrev "___tier" tier
  let (f',t',attrs) = 
        if order == Nothing
           then (f,t,[])
           else let (Just sls) = order
                    in case (findIndex (==from) sls, findIndex (==to) sls) of
                         (Just x, Just y) -> if x>y then (t,f,[("dir","back")]) else (f,t,[])
                         _                -> (f,t,[])

  edge f' t' $ [("label",l)] ++ attrs
  incTier
    
-- Order setting is done in Msg processing
flowElement2dot _ (Order _) = return ()



mkLabel :: String -> Diagram String
mkLabel lbl = do
  t <- gets numTier
  return $ show t ++ ": " ++ reflow lbl

invisNode :: Diagram D.NodeId
invisNode = node [("style","invis"),("shape","point")]

reflow :: String -> String
-- FIXME: for now, you have to hardcode desired width/height ratio
-- FIXME: (tail $ init $ show) trick is needed to work around dotgen-0.2 limitations
reflow str = intercalate "\n" $ map unwords $ splitInto words_in_row w
      where w = words str
            z = length w
            rows = z*height `div` (height+width)
            words_in_row = rows*width `div` height
            chunk _ []  = Nothing
            chunk 0 lst = Just (lst, [])
            chunk n lst = Just $ splitAt n lst
            splitInto n = unfoldr (chunk n)
            width=3
            height=1

-- Return the ID of the next node in the swimline `name',
-- generating all required nodes and swimline connections along the way
connectToPrev :: String -> D.NodeId -> Diagram ()
connectToPrev "___tier" _ = return ()
connectToPrev sline currNode = do
  s <- getSwimline sline
  case s of
       -- Swimline already exists
       (Just prevNode) ->  do edge prevNode currNode [("style","dotted"),("arrowhead","none")]
                              setSwimline sline currNode
       -- Otherwise, swimline hase to be created
       (Nothing) -> do setSwimline sline currNode
                       -- Add heading node
                       heading <- node [("label", mkHeader sline)]
                       addHeading heading
                       setSwimline sline heading
                       -- Retry connecting
                       connectToPrev sline currNode

mkHeader :: String -> String
mkHeader = map remove_underscore
  where
    remove_underscore '_' = ' '
    remove_underscore x   = x

------------------------------
-- State access/modify helpers
------------------------------

incTier :: Diagram ()
incTier = modify (\e -> e {numTier = numTier e +1} )

getSwimline :: String -> Diagram (Maybe D.NodeId)
getSwimline name = do
  s <- gets swimlines
  return $ M.lookup name s

setSwimline :: String -> D.NodeId -> Diagram ()
setSwimline name x = do
  modify (\e -> e {swimlines = M.insert name x (swimlines e)})

addHeading :: D.NodeId -> Diagram ()
addHeading x = do
  modify (\e -> e {headings = x:(headings e)})

------------------------------------------------
-- Lifting Text.Dot functions to the State monad
------------------------------------------------
same = lift . D.same
node = lift . D.node
edge f t args = lift $ D.edge f t args

---------
-- Parser
---------
-- | Parse specified file and return Flow Diagram contained therein.
-- All syntax errors are thrown via 'error'
parseFlowFromFile :: FilePath -> IO [Flow]
parseFlowFromFile fname = do
  raw <- readFile fname
  return $ parseFlow fname raw

-- | Parse given string and return Flow Diagram contained therein.
-- All syntax errors are thrown via 'error'
parseFlow :: String -> String -> [Flow]
parseFlow _     ""  = []
parseFlow fname str =
  case parse document fname str of
       Left err   -> error $ unlines [ "Input:", str, "Error:", show err]
       Right flow -> flow

document :: GenParser Char st [Flow]
document = do
  whitespace
  fl <- many flowLine
  eof
  return fl

flowLine, parseMsg, parseAction :: GenParser Char st Flow
flowLine = try parseOrder <|> try parseMsg <|> try parseAction
parseOrder = do string "order"
                is <- identifier `manyTill` newline
                return $ Order is
parseMsg = do f <- identifier; string "->"; t <- identifier; string ":"; m <- anything
              return $ Msg f t (trim m)
parseAction = do s <- identifier; string ":"; a <- anything
                 return $ Action s (trim a)

identifier, whitespace, anything :: GenParser Char st String
identifier = do whitespace; i <- many1 (alphaNum <|> oneOf "_"); whitespace
                return i
whitespace = many $ oneOf " \t"
anything = try (anyChar `manyTill` newline) <|> many1 anyChar

trim :: String -> String
trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace

-- Parser tests
newtype Name = Name String
newtype Message = Message String

instance Arbitrary Name where
  arbitrary = liftM Name (listOf' $ elements "abcxyz_банк")
  coarbitrary = undefined

instance Arbitrary Message where
  -- words.unwords trick is needed to prevent Messages which contain only spaces
  arbitrary = liftM ((Message).unwords.words) $ frequency [ (50, listOf' $ elements "abcxyz_->; 123банк")
                                                          -- One special case which i decided to hard-code
                                                          , (1, return "foo -> bar")
                                                          ]
  coarbitrary = undefined

instance Arbitrary Flow where
  arbitrary = frequency [ (10, liftM3 Msg mkName mkName mkMsg)
                        , (5, liftM2 Action mkName mkMsg)
                        ]
    where
      mkName = do Name n <- arbitrary; return n
      mkMsg = do Message m <- arbitrary; return m
  coarbitrary = undefined

-- Taken from a unreleased version of quickcheck
-- Just added ' to the names
--   / Kolmodin
listOf' :: Gen a -> Gen [a]
listOf' gen = sized $ \n ->
  do k <- choose (1,n)
     vectorOf' k gen

vectorOf' :: Int -> Gen a -> Gen [a]
vectorOf' k gen = sequence [ gen | _ <- [1..k] ]


-- | Print element of the flow diagram as String
showFlow :: Flow -> String
showFlow (Order sl)   = "order " ++ intercalate " " sl
showFlow (Msg f t m)  = unwords [ f, " -> ", t, ":", m ]
showFlow (Action s a) = unwords [ s, ":", a ]

prop_reparse :: [Flow] -> Bool
prop_reparse x =
  let txt = unlines $ map showFlow x
      in x == parseFlow "" txt

prop_russian_k :: Bool
prop_russian_k =
  ( parseFlow "a->b" "A->B: клиент" == [Msg "A" "B" "клиент"] ) &&
  ( parseFlow "prod" "продавец -> клиент: подписание контракта, предоставление счета" == [Msg "продавец" "клиент" "подписание контракта, предоставление счета"] )
